src/Http2Client.ts

```diff
@@ -40,38 +44,82 @@ export class Http2Client extends EventEmitter implements Client {
 	 *
 	 * @param host - hostname or IP address of daemon
 	 * @param port - port number of daemon on the host
-	 * @override
 	 */
-	introduce(host: string, port: number): void {
+	public introduce(host: string, port: number): void {
 		const client = connect(`http://${hostAndPort(host, port)}`)
 		client.on('error', (err) => this.emit('error', err))
-		this.available(client)
+		this.setAvailableAndCheckJobs(client)
 	}
 
 	/**
-	 * Drive the given daemon, by providing it with one task at a time in an async loop.
+	 * Mark the daemon as available.
+	 * Then calls function to check if there are any doable jobs.
 	 *
-	 * @param client - a HTTP/2 client connected to a daemon.
+	 * @param daemon - The daemon that is available.
+	 */
+	private setAvailableAndCheckJobs(daemon: ClientHttp2Session): void {
+		this.availableDaemons.add(daemon)
+		this.checkJobsAndAssign()
+	}
+
+	/**
+	 * Tries to assign the next available job to a daemon.
+	 * Closes all daemons if all jobs are finished.
 	 */
-	private available(client: ClientHttp2Session): void {
-		const job = this.queue.pop()
-		if (!job) {
-			client.close(() => this.emit('done'))
-			return
+	private checkJobsAndAssign(): void {
+		const job = this.jobOrderer.popNextJob()
+
+		if (job && this.availableDaemons.size > 0) {
+			const daemon: ClientHttp2Session = this.availableDaemons.values().next()
+				.value as ClientHttp2Session
+
+			assert(daemon, 'No available daemon found (logic error).')
+
+			this.assignJobToDaemon(job, daemon)
+		} else if (this.jobOrderer.isDone()) {
+			this.closeAllDaemonsAndFinish()
 		}
+	}
 
-		const req = client.request({ ':path': `/${job}` })
-		client.on('error', (err) => this.emit('error', err))
+	/**
+	 * Closes all daemons.
+	 * Emits 'done' once all daemons are closed.
+	 */
+	private closeAllDaemonsAndFinish() {
+		for (const daemon of this.availableDaemons) {
+			daemon.close(() => {
+				this.availableDaemons.delete(daemon)
+				if (this.availableDaemons.size === 0) {
+					this.emit('done')
+				}
+			})
+		}
+	}
+
+	/**
+	 * Asks the daemon to work on the job.
+	 * Sends HTTP request and handles response.
+	 *
+	 * @param job - The job to assign.
+	 * @param daemon - The daemon to which to assign the job.
+	 */
+	private assignJobToDaemon(job: Job, daemon: ClientHttp2Session) {
+		const request = daemon.request({ ':path': `/${job.getName()}` })
+		daemon.on('error', (err) => this.emit('error', err))
```

Comment:
Instead of emitting error, we can emit a less disastrous event (or none at all) and handle this within the Client. The error handler could
1. Tell JobOrderer to "reportInconclusiveJob" (I just made that up) which puts it back as if we had never tried it. Or have some other way of holding on to this job.
2. Implement some policy re. the daemon; in the simplest case we could just close the connection and remove the daemon from the list.
3. Call checkJobsAndAssign

Thoughts?

Reply:
After reading HeapJobOrderer, I realize this is already in place (all you need to do is call reportFailedJob from this error handler). No need to implement it in this PR.